raft-tech/TANF-app
# 2870 elastic reindex customizations

## Summary of Changes
Along with #2880 , pull request closes #2870 

* creates `tdp_seardch_index`, an override of the `django-elasticsearch-dsl` `manage.py` command `search_index` that allows passing in a couple options for `thread_count` and `chunk_size`, as well as settings for elastic's slow index and search query log thresholds.
* creates a celery beat scheduled task to perform `python manage.py tdp_search_index --rebuild --use-alias --parallel -f` ~~every Saturday at 1am EST~~
   * The beat schedule has been **REMOVED**, so the task does not run. 
* also moved the handy `log` function to `core.utils`

## How to Test
```
cd tdrs-backend && docker-compose up
cd tdrs-frontend && docker-compose up --build
```

**Note** Crontab has been disabled, see additional testing steps for the management command below
1. Ensure some parsing data is loaded into your postgres db.
1. Adjust the `schedule` for "Reindex elasticsearch documents" in `settings/common.py`. 
   - `crontab(minute='*', hour='*', day_of_week='*', day_of_month='*', month_of_year='*')`
1. There should be a django admin `LogEntry` created when reindexing starts and another when reindexing ends. There should also be a ton of output in the console when the reindex runs.


Test the `tdp_search_index` management command.
1. `cf ssh` into the deployed backend
1. Run `/tmp/lifecycle/shell` to enable the shell
1. Run `python manage.py tdp_search_index --rebuild`. All options from `search_index` should pass through (`-f`, `--use-alias`, `--models`, etc). Use environment variables to control thresholds and settings. 
1. Verify log thresholds are preserved. `cf ssh` into the elastic proxy instance. Run `curl http://localhost:8080/{index}/_settings` (replace `{index}` with any index name from `curl http://localhost:8080/_cat/indices?v`)

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [ ] cron/beat task to reindex the entire database periodically

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
+ Are code coverage minimums met?

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)


### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)



### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)


### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)


### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)


### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
